src/V3Param.cpp

```diff
@@ -718,7 +718,8 @@ class ParamProcessor final : public VNDeleter {
                                   << pinp->prettyNameQ() << " of " << nodep->prettyNameQ());
                     pinp->exprp()->replaceWith(new AstConst{
                         pinp->fileline(), AstConst::WidthedValue{}, modvarp->width(), 0});
-                } else if (origp && exprp->sameTree(origp)) {
+                } else if (origp
+                           && V3Hasher::uncachedHash(exprp) == V3Hasher::uncachedHash(origp)) {
```

Comment:
This will false match, things may hash to the same value but not be identical.  Maybe sameTree (and same() in each Ast) needs improvement for strings?

Reply:
The problem with `sameTree` is that is compares dtypep() pointers, which may be different, but these types may be the same. I think that that comparison can be replaced with `similarDType` method. I will check.

Reply:
It seems to work and I think it is correct. However, I wonder what to do if one node has a non-null dtypep() and the other doesn't. Currently it will return false but maybe it possible that two nodes represent the same value, but one was already handled by V3Width (because V3Const::constifyParamsEdit` was used) and the other wasn't. In such a case, the first one will probably have non-nullptr dtype(), but the other won't.

Reply:
I believe simiarDType is only for dtypes and here we want to compare the real values.   We could perhaps make a more generic similar that does what you need and replace similarDType with that - if so I'd prefer a separate pull request for that rename before doing this (it's ok to put here for testing then extract that code later to a pull).

I think we should have already constify'ed both sides of the compare at this point.  Maybe add an assertion that dtypep() is at least set on both sides before calling similar?

Reply:
I meant that the comparison of dtypep() pointers from `sameTreeIter` should be replaced by `similarDType`. I committed it already.

Reply:
I added such an assert (https://github.com/antmicro/verilator-1/commit/be5681e351e64b127bd72ed74c924a54cc65ec69) and many tests failed. I checked one of them and the `sameTree` method was called from https://github.com/verilator/verilator/blob/7005a65d32dfa86afa88234688001c686d8bf9b1/src/V3DupFinder.cpp#L53, from V3MergeCond stage. I think it is a separate problem that such cases occur.

Reply:
It started failing when I constify the default value of a parameter. I don't know what is the problem yet.

Reply:
Default values of parameters are not always constified. And I think that it is a separate task.
If there is a parameter with the default value referencing other parameters, e.g: `class Cls #(parameter int X =10, Y=2*X)` and X is overridden by the user and Y is not, we need to evaluate the value of `Y` by taking the actual value of X instead of the default. Such logic isn't implemented. If we add an `UNSUPPORTED` error in that case, we will probably broke some cases that currently works.

I think that for this issue we can leave the default values as they are now.